recipes_source/recipes/reasoning_about_shapes.py

```diff
@@ -0,0 +1,83 @@
+"""
+Reasoning about Shapes in PyTorch
+=================================
+
+When writing models with PyTorch, it is commonly the case that the parameters
+to a given layer depend on the shape of the output of the previous layer. For
+example, the ``in_features`` of an ``nn.Linear`` layer are dependent on
+``.size(-1)`` of the input. For some layers, the shape computation involves
+complex equations, for example convolution operations.
+
+One way around this is to run the forward pass with random inputs, but this is
+wasteful in terms of compute.
+
+Instead, we can make use of the ``meta`` device to determine the output shapes
+of a layer without materializing any data.
+"""
+
+import torch
+import timeit
+
+start = timeit.default_timer()
+t = torch.rand(2, 3, 10, 10, device="meta")
+end = timeit.default_timer()
+conv = torch.nn.Conv2d(3, 5, 2, device="meta")
+
+print(conv(t))
+print(f"Time taken: {end-start}")
+
+
+##########################################################################
+# Observe that since data is not materialized, passing arbitrarily large
+# inputs will not significantly alter the time taken for shape computation.
+
+start = timeit.default_timer()
+t_large = torch.rand(2**10, 3, 2**16, 2**16, device="meta")
+end = timeit.default_timer()
+
+print(conv(t_large))
+print(f"Time taken: {end-start}")
+
+######################################################
+# Consider an aribtrary network such as the following:
+
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class Net(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.conv1 = nn.Conv2d(3, 6, 5)
+        self.pool = nn.MaxPool2d(2, 2)
+        self.conv2 = nn.Conv2d(6, 16, 5)
+        self.fc1 = nn.Linear(16 * 5 * 5, 120)
+        self.fc2 = nn.Linear(120, 84)
+        self.fc3 = nn.Linear(84, 10)
+
+    def forward(self, x):
+        x = self.pool(F.relu(self.conv1(x)))
+        x = self.pool(F.relu(self.conv2(x)))
+        x = torch.flatten(x, 1) # flatten all dimensions except batch
+        x = F.relu(self.fc1(x))
+        x = F.relu(self.fc2(x))
+        x = self.fc3(x)
+        return x
+
+
+with torch.device("meta"):
+  net = Net()
```

Comment:
```suggestion
    net = Net()
```
(to match previous indentation)